File: Data_Structures/C++/BinarySearchTree/BinarySearchTree.cpp
#include <bits/stdc++.h>

using namespace std;

struct Node {
    int data;
    Node *left;
    Node *right;
};

Node    *newNode(int data)
{
    Node *new_node = new Node();
    new_node->data = data;
    new_node->left = NULL;
    new_node->right = NULL;
    return new_node;
}

Node    *Insert(Node *root, int data)
{
    if (!root)
        root = newNode(data);
    else if (root->data >= data)
        root->left = Insert(root->left, data);
    else
        root->right = Insert(root->right, data);
    return root;
}

Node    *Search(Node *root, int data)
{
    if (!root) return NULL;
    else if (root->data == data)
        return root;
    else if (root->data <= data)
        return Search(root->right, data);
    else
        return Search(root->left, data);
}

int main() {
	Node *root = NULL;
	/*Code to test the logic*/
	root = Insert(root, 15);
	root = Insert(root, 10);
	root = Insert(root, 20);
	root = Insert(root, 25);
	root = Insert(root, 8);
	root = Insert(root, 12);
	// Ask user to enter a number.
	int number;
	cout << "Enter number be searched\n";
	cin >> number;
	// If number is found, print "FOUND"
	if (Search(root, number)) cout << "Found\n";
	else cout << "Not Found\n";
}

File: Data_Structures/javascript/BinarySearchTree/BinarySearchTree.js
// Node
class Node {
    constructor(key) {
        this.key = key,
        this.left = null,
        this.right = null
    }
}

// BST
class BST {
    constructor() {
        this.root = null;
    }

    insertNode = (node, newNode) => {
        if (newNode.key < node.key) {
            if (node.left === null)
                node.left = newNode;
            else
                this.insertNode(node.left, newNode);
        } else {
            if (node.right === null)
                node.right = newNode;
            else
                this.insertNode(node.right, newNode);
        }
    }

    // Insert
    insert = key => {
        let newNode = new Node(key);

        if (this.root == null)
            this.root = newNode;
        else
            this.insertNode(this.root, newNode);
    }

    // Search
    search = (key, node = this.root) => {
        if (node === null)
            return false;

        if (key < node.key)
            return this.search(key, node.left);
        else if (key > node.key)
            return this.search(key, node.right);
        else
            return true;
    }

    // Min
    min = (node = this.root) => {
        if (node) {
            while (node && node.left !== null)
                node = node.left;
            return node.key;
        }
        return null;
    }

    // Max
    max = (node = this.root) => {
        if (node) {
            while (node && node.right !== null)
                node = node.right;
            return node.key
        }
        return null;
    }

    // Remove
    remove = (key) => {
        this.root = this.removeNode(this.root, key);
    }

    removeNode = (node, key) => {
        if (node === null)
            return null;

        if (key < node.key) {
            node.left = this.removeNode(node.left, key);
            return node;
        } else if (key > node.key) {
            node.right = this.removeNode(node.right, key);
            return node;
        } else {
            if (node.left === null && node.right === null) {
                node = null;
                return node;
            }

            if (node.left === null) {
                node = node.right;
                return node;
            } else if (node.right === null) {
                node = node.left;
                return node;
            }

            let aux = this.min(node.right);
            node.key = aux.key;
            node.right = this.removeNode(node.right, aux.key);
            return node;
        }
    }
}

const tree = new BST();
tree.insert(11);
tree.insert(7);
tree.insert(15);
tree.insert(5);
tree.insert(3);
tree.insert(9);
tree.insert(8);
tree.insert(10);
tree.insert(13);
tree.insert(12);
tree.insert(14);
tree.insert(20);
tree.insert(18);
tree.insert(25);

tree.remove(18);
console.log(tree.min());
console.log(tree.max());
console.log(tree.search(18));
